### PR TITLE
Improvements to map unification

### LINKED ISSUE #319: Booster aborts on rules fetching information from a concrete `Map` with multiple keys
@virgil-serbanuta has rules of the following form in https://github.com/runtimeverification/wasm-semantics:

```
rule <k> call(...) => stuff(V1, V2, V3) ... </k>
        <locals> 1 |-> V1 2 |-> V2 3 |-> V3 </locals>
```

i.e. matching on a cell which contains a `Map` with only concrete keys, and using the values for something else. These rules cause Booster to abort due to unification failure (we do not have the exact message).

### PR DESCRIPTION
**Pull request by goodlyrottenapple**
This fixes #319 and hopefully also #298 by newly allowing unification of:
* Two maps with multiple fully concrete keys and no "rest" map variable (previously we could only match multiple keys if there was also a "rest of the map" variable)
* Map with only concrete keys + "rest" variable with a map of mixed concrete/symbolic keys without a "rest" variable

### CODE REVIEW

**Comment by goodlyrottenapple** [comment]
Kontrol performance:
```
 Test                                                                                                                      | sam-map-unification-improvements time | master time | (sam-map-unification-improvements/master) time
---------------------------------------------------------------------------------------------------------------------------+---------------------------------------+-------------+------------------------------------------------
 src/tests/integration/test_foundry_prove.py::test_foundry_prove[AddrTest.test_notBuiltinAddress_concrete()]               | 55.89                                 | 53.95       | 1.0359592215013902
 src/tests/integration/test_foundry_prove.py::test_foundry_prove[PlainPrankTest.test_startPrankWithOrigin_true()]          | 211.98                                | 204.39      | 1.0371348891824455
 src/tests/integration/test_foundry_prove.py::test_foundry_prove[ExpectRevertTest.testFail_expectRevert_empty()]           | 76.69                                 | 73.88       | 1.0380346507850569
 src/tests/integration/test_foundry_prove.py::test_foundry_prove[ExpectRevertTest.test_expectRevert_message()]             | 151.49                                | 145.89      | 1.0383850846528209
 src/tests/integration/test_foundry_prove.py::test_foundry_fail[AssertTest.test_revert_branch(uint256,uint256)]            | 114.17                                | 109.12      | 1.0462793255131964
 src/tests/integration/test_foundry_prove.py::test_foundry_prove[ArithmeticTest.test_max1(uint256,uint256)]                | 123.0                                 | 117.47      | 1.0470758491529752
 src/tests/integration/test_foundry_prove.py::test_foundry_prove[LoopsTest.test_sum_10()]                                  | 96.54                                 | 91.73       | 1.052436498419274
 src/tests/integration/test_foundry_prove.py::test_foundry_prove[EmitContractTest.testExpectEmitCheckEmitter()]            | 143.76                                | 135.58      | 1.060333382504794
 src/tests/integration/test_foundry_prove.py::test_foundry_prove[AddrTest.test_addr_true()]                                | 66.52                                 | 62.15       | 1.0703137570394208
 src/tests/integration/test_foundry_prove.py::test_foundry_prove[ExpectRevertTest.testFail_ExpectRevert_failAndSuccess()]  | 163.32                                | 151.16      | 1.0804445620534533
 src/tests/integration/test_foundry_prove.py::test_foundry_prove[ExpectRevertTest.testFail_expectRevert_multipleReverts()] | 179.1                                 | 164.95      | 1.085783570779024
 src/tests/integration/test_foundry_prove.py::test_foundry_prove[BMCLoopsTest.test_countdown_concrete()]                   | 62.16                                 | 56.58       | 1.0986214209968186
 src/tests/integration/test_solc_to_k.py::test_solc_to_k[Empty.sol]                                                        | 55.9                                  | 50.21       | 1.1133240390360486
 src/tests/integration/test_solc_to_k.py::test_solc_to_k[ERC721.sol]                                                       | 56.53                                 | 48.85       | 1.1572159672466735
 TOTAL                                                                                                                     | 1557.05                               | 1465.91     | 1.062172984698924
```

**Comment by goodlyrottenapple** [comment]
Kevm performance:
```
 Test                                                                                      | sam-map-unification-improvements time | master time        | (sam-map-unification-improvements/master) time
-------------------------------------------------------------------------------------------+---------------------------------------+--------------------+------------------------------------------------
 src/tests/integration/test_prove.py::test_pyk_prove[erc20/hkg/approve-spec.k]             | 70.66                                 | 76.6               | 0.9224543080939948
 src/tests/integration/test_prove.py::test_pyk_prove[erc20/ds/transfer-failure-1-b-spec.k] | 169.29                                | 160.45             | 1.055095045185416
 TOTAL                                                                                     | 239.95                                | 237.04999999999998 | 1.0122337059692048
```

**Review by jberthold — APPROVED** [review]
LGTM
I would think the slowdown in the tests is acceptable since this will unlock elrond use cases and fix the problem reported in #298. Ideally, that proof, and an exemplary elrond proof, should be extracted as tarballs in a future regression test suite.

File: unit-tests/Test/Booster/Pattern/Unify.hs [hunk 1]
```diff
@@ -366,6 +367,77 @@ internalLists =
...
+            symbolicKMapWithTwoItems
+            ( success
+                [ ("A", kmapElementSort, [trm| \dv{SortTestKMapItem{}}("value")|])
+                , ("B", kmapElementSort, [trm| \dv{SortTestKMapItem{}}("value2")|])
+                ]
+            )
+        , test
+            "Can unify {\"key\" |-> \"value\", ...REST} with {A |-> \"value\"}"
+            concreteKMapWithOneItemAndRest
+            symbolicKMapWithOneItem
```
[the remark this answers is not in the document]
Comment (goodlyrottenapple) [reply unit-tests/Test/Booster/Pattern/Unify.hs:401]: ah good point, i will re-order

**Review by geo2a — APPROVED** [review]

**Pull request merged**